tools/include/eosio/abigen.hpp

```diff
@@ -530,5 +626,143 @@ namespace eosio { namespace cdt {
          std::set<abi_table>                   ctables;
          std::map<std::string, std::string>    rcs;
          std::set<const clang::Type*>          evaluated;
+
+         bool is_kv_table(const clang::CXXRecordDecl* decl) {
+            for (const auto& base : decl->bases()) {
+               auto type = base.getType();
+               if (type.getAsString().find("eosio::kv_table<") != std::string::npos) {
+                  return true;
+               }
+            }
+
+            return false;
+         }
+
+         std::string get_value_from_decltype(const clang::DecltypeType* decl) {
+            if (const auto ref_type = dyn_cast<clang::LValueReferenceType>(decl->desugar())) {
+               const auto pt = ref_type->getPointeeType();
+               if (const auto record_type = dyn_cast<clang::RecordType>(pt)) {
+                  const auto gdt = record_type->getDecl();
+                  if (const auto ctsd = dyn_cast<clang::ClassTemplateSpecializationDecl>(gdt)) {
+                     const auto& template_args = ctsd->getTemplateArgs();
+                     // Handle tuples, which have the template arguments treated as a pack
+                     if (template_args.size() == 1 && template_args[0].pack_size() > 1) {
+                        std::string ret = ctsd->getName().str();
+                        for (const auto& pack_elem : template_args[0].getPackAsArray()) {
+                           ret += "_";
+                           ret += get_type(pack_elem.getAsType());
+                        }
+                        return ret;
+                     } else {
+                        // Handle non-tuple templates
+                        return get_type(record_type->desugar());
+                     }
+                  } else {
+                     // Handle non-template records
+                     return get_type(record_type->desugar());
+                  }
+               } else {
+                  // Handle non-records
+                  return get_type(pt);
+               }
+            }
+            CDT_INTERNAL_ERROR("Error generating type from kv index definition");
+            return "";
+         }
+
+         const std::set<std::string> internal_types {
```

Comment:
These types should be namespace qualified.  It seems like this could cause issues with user defined types being flagged incorrectly.